### DISCUSSION TITLE
About Corrupt JPEG data

### DISCUSSION THREAD

**Discussion by jishi92** [libvips/libvips#3289] [Q&A]
Hi:

I have some corrupt JPEG imgs, some of them can be thumbnail and some can't and vips fail_on is none.

for example:
![corrupt-test](https://user-images.githubusercontent.com/22196414/213389906-a6303860-5a42-47a7-bd28-944fbb1390b1.jpeg)

It works in libvips 8.12.2 and doesn't work in 8.13.

```
➜  Downloads vipsthumbnail corrupt-test.jpeg --size=100x100

(vipsthumbnail:96427): VIPS-WARNING **: 16:10:21.693: read gave 118 warnings

(vipsthumbnail:96427): VIPS-WARNING **: 16:10:21.694: VipsJpeg: Corrupt JPEG data: premature end of data segment

➜  Downloads echo $?
0
➜  Downloads vips -v
vips-8.12.2-Tue Jan 25 09:34:32 UTC 2022
```

```
[root@3a68b95ade7b Downloads]# vipsthumbnail corrupt-test.jpeg --size=100x100

(vipsthumbnail:3031): VIPS-WARNING **: 16:09:00.541: error in tile 0 x 208

(vipsthumbnail:3031): VIPS-WARNING **: 16:09:00.541: error in tile 0 x 70
vipsthumbnail: unable to thumbnail corrupt-test.jpeg
VipsJpeg: Corrupt JPEG data: premature end of data segment
VipsJpeg: Corrupt JPEG data: found marker 0xd9 instead of RST1

(vipsthumbnail:3031): VIPS-WARNING **: 16:09:00.544: read gave 100 warnings

(vipsthumbnail:3031): VIPS-WARNING **: 16:09:00.544: 
[root@3a68b95ade7b Downloads]# echo $?
255
[root@3a68b95ade7b Downloads]# vips -v
vips-8.13.0
```

What should I do can make it work in 8.13?

**Comment by jcupitt** [comment]
Hi again,

I found an improvement in the jpeg loader and made a PR: https://github.com/libvips/libvips/pull/3293

I now see:

```
$ vips copy trunc4.jpg x.png

(vips:290658): VIPS-WARNING **: 13:17:36.519: read gave 100 warnings

(vips:290658): VIPS-WARNING **: 13:17:36.519: VipsJpeg: Corrupt JPEG data: premature end of data segment
VipsJpeg: Corrupt JPEG data: found marker 0xd9 instead of RST1
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848

memory: high-water mark 5.26 MB
error buffer: VipsJpeg: Corrupt JPEG data: premature end of data segment
VipsJpeg: Corrupt JPEG data: found marker 0xd9 instead of RST1
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
VipsJpeg: out of order read at line 848
$ echo $?
0
```

So a lot of errors and warnings, but it does return 0. Any testing very welcome, if you're able.

**Reply by jishi92** [reply]
Hi @jcupitt,
Is there any plan to merge the PR: https://github.com/libvips/libvips/pull/3293?
I have been troubled by the corrupt image for a long time, but I don't want to go back to version 8.12, for fear of some unknown problems.

**Comment by jishi92** [comment]
Oh,thank you. 
You're so quick.
So this problem will be fix in latest version ?

**Reply by jcupitt** [reply]
Sorry, no idea, It'll depend how it gets merged.

**Reply by jishi92** [reply]
OK, I know.
By the way, Is there any func can judge if the image is truncated?

**Reply by jcupitt** [reply]
It's not possible to tell in all cases. Every load library handles this differently, and does not always expose all the information you need. 

libvips tries to detect truncation as well as it can, but it's not always perfect.

**Reply by jishi92** [reply]
What about the most common jpg and png?
Could you consider providing a func to detect if the image(just png and jpeg) is truncated?

**Reply by jcupitt** [reply]
Yes, libvips tries to do this, but libjpeg and libpng don't always make this possible.

**Comment by kleisauke** [comment]
Looking at this carefully, it appears that the corrupted image triggers more than 100 warnings, leading to this failure. This behavior was introduced in commit df26bd1e46c67ed7745996627ca0d83e5a6d678d. 

Commit 2c4c0390564da4acf4186663b6bcd1c4ccbe96d8 adds an `unlimited` flag to allow these images to be processed. So, you can "fix" this by doing:
```console
$ vipsthumbnail corrupt-test.jpg[unlimited] -s 100x100

(vipsthumbnail:112858): VIPS-WARNING **: 11:13:44.619: read gave 118 warnings

(vipsthumbnail:112858): VIPS-WARNING **: 11:13:44.619: VipsJpeg: Corrupt JPEG data: premature end of data segment

$ vipsheader tn_corrupt-test.jpg 
tn_corrupt-test.jpg: 75x100 uchar, 3 bands, srgb, jpegload
```

Note to self: commit df26bd1e46c67ed7745996627ca0d83e5a6d678d can be reverted and the `unlimited` flag could be deprecated after libjpeg-turbo v3.0.4 is released, see: https://github.com/libjpeg-turbo/libjpeg-turbo/issues/764. However, we should probably wait until MozJPEG also incorporates this.

**Reply by jcupitt** [reply]
Clever, I hadn't thought of that.
